server/src/server.rs

```diff
@@ -36,26 +37,161 @@ fn error(error: cubeb::Error) -> ClientMessage {
     ClientMessage::Error(error.raw_code())
 }
 
-type ContextKey = RefCell<Option<cubeb::Result<cubeb::Context>>>;
-thread_local!(static CONTEXT_KEY: ContextKey = RefCell::new(None));
+struct CubebDeviceCollectionManager {
+    servers: Vec<Rc<RefCell<CubebServerCallbacks>>>,
+    input_registered: bool,
+    output_registered: bool,
+}
+
+impl CubebDeviceCollectionManager {
+    fn new() -> CubebDeviceCollectionManager {
+        CubebDeviceCollectionManager {
+            servers: Vec::new(),
+            input_registered: false,
+            output_registered: false,
+        }
+    }
+
+    fn register(&mut self, context: &cubeb::Context, server: &Rc<RefCell<CubebServerCallbacks>>) {
+        if self
+            .servers
+            .iter()
+            .find(|s| Rc::ptr_eq(s, server))
+            .is_none()
+        {
+            self.servers.push(server.clone());
+        }
+        self.update(context);
+    }
+
+    fn unregister(&mut self, context: &cubeb::Context, server: &Rc<RefCell<CubebServerCallbacks>>) {
+        self.servers
+            .retain(|s| !(Rc::ptr_eq(&s, server) && s.borrow().devtype.is_empty()));
```

Comment:
Why using `retain` instead of `remove_item`?

Reply:
remove_item is unstable (nightly only).